core/src/index/knowhere/knowhere/index/structured_index/StructuredIndexSort-inl.h

```diff
@@ -0,0 +1,224 @@
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software distributed under the License
+// is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+// or implied. See the License for the specific language governing permissions and limitations under the License
+
+#include <algorithm>
+#include <memory>
+#include <utility>
+#include "knowhere/index/structured_index/StructuredIndexSort.h"
+
+namespace milvus {
+namespace knowhere {
+
+template <typename T>
+StructuredIndexSort<T>::StructuredIndexSort() : is_built_(false), data_(nullptr), n_(0) {
+}
+
+template <typename T>
+StructuredIndexSort<T>::StructuredIndexSort(const size_t n, const T* values) : is_built_(false), n_(n) {
+    Build(n, values);
+}
+
+template <typename T>
+StructuredIndexSort<T>::~StructuredIndexSort() {
+}
+
+template <typename T>
+void
+StructuredIndexSort<T>::Build(const size_t n, const T* values) {
+    data_.reserve(n);
+    T* p = const_cast<T*>(values);
+    for (auto i = 0; i < n; ++i) {
+        //        data_[i].a_ = *p ++;
+        //        data_[i].idx_ = i;
+        data_.emplace_back(IndexStructure(*p++, i));
+    }
+    build();
+}
+
+template <typename T>
+void
+StructuredIndexSort<T>::build() {
+    if (is_built_)
+        return;
+    if (data_.size() == 0 || n_ == 0) {
+        // todo: throw an exception
+        KNOWHERE_THROW_MSG("StructuredIndexSort cannot build null values!");
+    }
+    std::sort(data_.begin(), data_.end());
+    is_built_ = true;
+}
+
+template <typename T>
+BinarySet
+StructuredIndexSort<T>::Serialize(const milvus::knowhere::Config& config) {
+    if (!is_built_) {
+        build();
+    }
+
+    auto index_data_size = n_ * sizeof(IndexStructure<T>);
+    std::shared_ptr<uint8_t[]> index_data(new uint8_t[index_data_size]);
+    memcpy(index_data.get(), data_.data(), index_data_size);
+
+    std::shared_ptr<uint8_t[]> index_length(new uint8_t[sizeof(size_t)]);
+    memcpy(index_length.get(), &n_, sizeof(size_t));
+
+    BinarySet res_set;
+    res_set.Append("index_data", index_data, index_data_size);
+    res_set.Append("index_length", index_length, sizeof(size_t));
+    return res_set;
+}
+
+template <typename T>
+void
+StructuredIndexSort<T>::Load(const milvus::knowhere::BinarySet& index_binary) {
+    try {
+        auto index_length = index_binary.GetByName("index_length");
+        memcpy(&n_, index_length->data.get(), (size_t)index_length->size);
+
+        auto index_data = index_binary.GetByName("index_data");
+        data_.reserve(n_);
+        memcpy(data_.data(), index_data->data.get(), (size_t)index_data->size);
+        is_built_ = true;
+    } catch (...) {
+        KNOHWERE_ERROR_MSG("StructuredIndexSort Load failed!");
+    }
+}
+
+// find the first element's offset which is no less than given value
+template <typename T>
+size_t
+StructuredIndexSort<T>::lower_bound(const T& value) {
```

Comment:
Why not use std::lower_bound and up_bound, since the container data_ is a vector?